### PR TITLE
Sorting

### PR DESCRIPTION
**Pull request by WeszeRita**

### CODE REVIEW

File: src/app/pages/todos/components/todo-card/todo-card.component.html [hunk 1]
```diff
@@ -1,7 +1,12 @@
 <form class="form">
   <span class="label">{{ buildTranslationKey('status') | translate }}</span>
 
-  <app-dropdown [formControl]="statusControl" [status]="todo.status"></app-dropdown>
+  <app-dropdown
+    [formControl]="statusControl"
+    [name]="buildTranslationKey(todo.status)"
+    [options]="options"
+    [statusClass]="statusClass">
```
Comment (bodnya29179) [inline src/app/pages/todos/components/todo-card/todo-card.component.html]: now your dropdown component should be generic and shouldn't depend on some status

File: src/shared/components/dropdown/dropdown.component.ts [hunk 2]
```diff
@@ -11,31 +12,38 @@ import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
...
       useExisting: DropdownComponent,
       multi: true,
-    }
+    },
   ],
 })
 export class DropdownComponent implements ControlValueAccessor {
-  statusOptions: string[] = Object.keys(Status);
+  @Input()
+  value: Status | SortTerm;
```
Comment (bodnya29179) [inline src/shared/components/dropdown/dropdown.component.ts]: value type should be also reusable, for example `string`

File: src/shared/components/dropdown/dropdown.component.ts [hunk 3]
```diff
@@ -11,31 +12,38 @@ import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
...
+
+  @Input()
+  name: IOption['value'];
+
+  @Input()
+  statusClass: Status;
```
Comment (bodnya29179) [inline src/shared/components/dropdown/dropdown.component.ts]: As I mentioned, it shouldn't be inside the dropdown component

File: src/shared/components/dropdown/dropdown.component.ts [hunk 4]
```diff
@@ -11,31 +12,38 @@ import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
...
-
-  setStatus(status: string): void {
-    if (this.status !== status) {
-      this.writeValue(status);
+  setStatus(value: Status | SortTerm, translatedValue: string): void {
+    if (this.value !== value) {
+      this.writeValue(value);
     }
+    this.name = translatedValue;
   }
```
Comment (bodnya29179) [inline src/shared/components/dropdown/dropdown.component.ts]: your displayed text should be set by getting an option from the options array by selected optionId

File: src/shared/components/header/header.component.html [hunk 5]
```diff
@@ -3,13 +3,20 @@
...
 
 <div class="search-sort">
-  <app-search></app-search>
+  <app-search [placeholder]="buildTranslationKey('searchPlaceholder') | translate"></app-search>
+
+  <app-dropdown
+    [formControl]="sortControl"
+    [icon]="iconSrc"
+    [options]="options"
+    [name]="options[0].value">
```
Comment (bodnya29179) [inline src/shared/components/header/header.component.html]: it's better to pass [selectedOption]="options[0]" and then you can do manipulations of settings values inside the dropdown component

File: src/shared/components/header/header.component.ts [hunk 6]
```diff
@@ -1,15 +1,56 @@
...
+    },
+    {
+      id: SortTerm.status,
+      value: `${ this.buildTranslationKey(SortTerm.status) }`,
+    },
+    {
+      id: SortTerm.name,
+      value: `${ this.buildTranslationKey(SortTerm.name) }`,
+    },
+  ];
```
Comment (bodnya29179) [inline src/shared/components/header/header.component.ts]: translate this values

File: src/shared/components/header/header.component.ts [hunk 7]
```diff
@@ -1,15 +1,56 @@
...
+
+  iconSrc = 'assets/icons/sort.svg';
+  sortControl: FormControl<SortTerm>;
+
+  constructor(public dialog: Dialog, private sortFacadeService: SortFacadeService, private destroyRef: DestroyRef) {}
```
Comment (bodnya29179) [inline src/shared/components/header/header.component.ts]: why is it public?

File: src/shared/components/header/header.component.ts [hunk 8]
```diff
@@ -1,15 +1,56 @@
...
+      id: SortTerm.status,
+      value: `${ this.buildTranslationKey(SortTerm.status) }`,
+    },
+    {
+      id: SortTerm.name,
+      value: `${ this.buildTranslationKey(SortTerm.name) }`,
+    },
+  ];
+
+  iconSrc = 'assets/icons/sort.svg';
```
Comment (bodnya29179) [inline src/shared/components/header/header.component.ts]: unused variable

File: src/shared/components/header/header.component.ts [hunk 9]
```diff
@@ -1,15 +1,56 @@
...
+      .pipe(
+        distinctUntilChanged(),
+        takeUntilDestroyed(this.destroyRef),
+      )
+      .subscribe((value: SortTerm) => {
+        this.options.filter((option) => {
+          if (option.id === value) {
+            this.sortFacadeService.sort(option.id);
+          }
+        });
```
Comment (bodnya29179) [inline src/shared/components/header/header.component.ts]: filter(), are you sure?

File: src/shared/components/header/header.component.ts [hunk 10]
```diff
@@ -1,15 +1,56 @@
...
+  constructor(public dialog: Dialog, private sortFacadeService: SortFacadeService, private destroyRef: DestroyRef) {}
+
+  ngOnInit(): void {
+    this.sortControl = new FormControl<SortTerm>(SortTerm.date);
+
+    this.sortControl.valueChanges
+      .pipe(
+        distinctUntilChanged(),
+        takeUntilDestroyed(this.destroyRef),
+      )
```
Comment (bodnya29179) [inline src/shared/components/header/header.component.ts:50]: make the CVA implementation to have the option object inside control for not to search it each time

File: src/shared/enums/sortTerm.enum.ts [hunk 11]
```diff
@@ -0,0 +1,5 @@
+export enum SortTerm {
```
Comment (bodnya29179) [inline src/shared/enums/sortTerm.enum.ts]: SortOption

File: src/shared/utils/search.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import { ITodo } from '@shared';
+
+export function search(array: ITodo[], searchTerm: string) {
```
Comment (bodnya29179) [inline src/shared/utils/search.ts]: the property name shouldn't be as the type.
it's todos

Comment (bodnya29179) [inline src/shared/utils/search.ts]: searchValue

File: src/shared/utils/search.ts [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+  if (!searchTerm) {
+    return array;
+  }
+
+  return array.filter((item: ITodo) => {
```
Comment (bodnya29179) [inline src/shared/utils/search.ts]: rename item to todo

File: src/shared/utils/sort.ts [hunk 14]
```diff
@@ -0,0 +1,24 @@
+import { IOption, ITodo } from '../interfaces';
+import { SortTerm,IStatusOrder } from '../enums';
+
+export function sort(array: ITodo[], sortTermId: IOption['id']) {
```
Comment (bodnya29179) [inline src/shared/utils/sort.ts]: the same about array
and `sortTermId` should be `sortOption: SortTerm`

File: src/shared/utils/sort.ts [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+        return IStatusOrder[a.status] - IStatusOrder[b.status];
+      });
+
+    case SortTerm.name:
+      return [...array].sort((a: ITodo, b: ITodo) => {
+        return a.title.localeCompare(b.title);
+      });
+
+    default:
+      return [...array];
```
Comment (bodnya29179) [inline src/shared/utils/sort.ts]: you don't need to copy array here

File: src/shared/components/header/header.component.ts (hunk 8 above)
Comment (WeszeRita) [reply src/shared/components/header/header.component.ts:33]: it is not unused, I use it to display the icon inside sort drop down

Comment (WeszeRita) [reply src/shared/components/header/header.component.ts]: Please check the solution in dropdown.component.html, I am not sure about it.

File: src/shared/components/dropdown/dropdown.component.html [hunk 16]
```diff
@@ -1,13 +1,17 @@
-<button class="button-title {{ statusClass }}" [cdkMenuTriggerFor]="dropdown"><img src="{{ icon }}">{{ name | translate }}</button>
+<button
+  class="button-title {{ initialValue }}"
```
Comment (bodnya29179) [inline src/shared/components/dropdown/dropdown.component.html]: why is initialValue inside class?

File: src/shared/components/dropdown/dropdown.component.html [hunk 17]
```diff
@@ -1,13 +1,17 @@
...
+  [cdkMenuTriggerFor]="dropdown">
+  <img src="{{ icon }}">{{ initialValue }}
```
Comment (bodnya29179) [inline src/shared/components/dropdown/dropdown.component.html]: don't use interpolation here, use simple data-binding

File: src/shared/components/dropdown/dropdown.component.ts [hunk 18]
```diff
@@ -17,33 +16,37 @@ import { IOption } from '../../interfaces/option.interface';
...
+    if (this.initialValue !== value) {
       this.writeValue(value);
     }
-    this.name = translatedValue;
+    this.options.map((option: IOption) => {
+      if (option.id === value) {
+        this.selectedOption = option;
+      }
+    });
+    this.placeholder =  this.selectedOption.value || this.initialValue;
```
Comment (bodnya29179) [inline src/shared/components/dropdown/dropdown.component.ts]: extra space

File: src/shared/components/dropdown/dropdown.component.ts [hunk 19]
```diff
@@ -17,33 +16,37 @@ import { IOption } from '../../interfaces/option.interface';
...
-  @Input()
-  options: IOption[];
+  placeholder: string;
 
   private onChange: (value: string) => void = () => {};
   private onTouched: () => void = () => {};
 
-  setStatus(value: Status | SortTerm, translatedValue: string): void {
-    if (this.value !== value) {
+  setStatus(value: string): void {
```
Comment (bodnya29179) [inline src/shared/components/dropdown/dropdown.component.ts]: why the method in the generic component has the name "status"?

File: src/shared/components/dropdown/dropdown.component.html [hunk 20]
```diff
@@ -1,13 +1,17 @@
...
 <ng-template #dropdown>
-  <div class="dropdown" cdkMenu>
+  <div cdkMenu class="dropdown">
     <button
       *ngFor="let option of options"
       class="option-item"
       cdkMenuItem
-      (click)="setStatus(option.id, option.value)">
-      {{ option.value | translate }}
+      (click)="setStatus(option.id)">
```
Comment (bodnya29179) [inline src/shared/components/dropdown/dropdown.component.html]: you don't need to pass option id and then search this option in list to select selectedOption.
Pass the whole option to method

File: src/app/pages/todos/components/todo-card/todo-card.component.scss [hunk 21]
```diff
@@ -13,3 +13,15 @@
...
+  @apply border border-orange-500 bg-orange-200 text-orange-500;
+}
+
+::ng-deep .todo button {
+  @apply border border-red-500 bg-red-200 text-red-500;
+}
+
+::ng-deep .done button {
+  @apply text-cyan-500 border bg-cyan-200 border-cyan-500;
+}
```
Comment (bodnya29179) [inline src/app/pages/todos/components/todo-card/todo-card.component.scss:40]: 1. repeating selector ng-deep.
2. wrap ng-deep into :host

File: src/app/pages/todos/components/todo-card/todo-card.component.ts [hunk 22]
```diff
@@ -33,7 +33,8 @@ export class TodoCardComponent implements OnInit {
   constructor(private todoFacadeService: TodoFacadeService, private destroyRef: DestroyRef, private translateService: TranslateService) {}
 
   ngOnInit(): void {
-    this.statusControl = new FormControl<Status>(this.todo.status);
+    const todo = this.options.find((option) => option.id === this.todo.status);
```
Comment (bodnya29179) [inline src/app/pages/todos/components/todo-card/todo-card.component.ts]: rename todo to status

File: src/shared/components/dropdown/dropdown.component.html [hunk 23]
```diff
@@ -1,7 +1,7 @@
 <button
-  class="button-title {{ initialValue }}"
+  class="button"
   [cdkMenuTriggerFor]="dropdown">
-  <img src="{{ icon }}">{{ initialValue }}
+  <img [src]="icon">{{ value }}
```
Comment (bodnya29179) [inline src/shared/components/dropdown/dropdown.component.html]: add class 'icon' to img and use in css

File: src/shared/components/header/header.component.scss [hunk 24]
```diff
@@ -13,3 +13,7 @@
 .btn {
   @apply bg-blue-600 rounded text-white px-1.5 py-1;
 }
+
+::ng-deep button[aria-expanded="true"] {
```
Comment (bodnya29179) [inline src/shared/components/header/header.component.scss]: the same about global styles

Comment (bodnya29179) [reply src/shared/components/header/header.component.scss]: handle it inside the component by using HostBinding

**Review by bodnya29179 — APPROVED** [review]

**Pull request merged**